test/core/transport/call_filters_test.cc

```diff
@@ -0,0 +1,1466 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/lib/transport/call_filters.h"
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::Mock;
+using testing::StrictMock;
+
+namespace grpc_core {
+
+namespace {
+// Offset a void pointer by a given amount
+void* Offset(void* base, size_t amt) { return static_cast<char*>(base) + amt; }
+
+// A mock activity that can be activated and deactivated.
+class MockActivity : public Activity, public Wakeable {
+ public:
+  MOCK_METHOD(void, WakeupRequested, ());
+
+  void ForceImmediateRepoll(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void Orphan() override {}
+  Waker MakeOwningWaker() override { return Waker(this, 0); }
+  Waker MakeNonOwningWaker() override { return Waker(this, 0); }
+  void Wakeup(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void WakeupAsync(WakeupMask /*mask*/) override { WakeupRequested(); }
+  void Drop(WakeupMask /*mask*/) override {}
+  std::string DebugTag() const override { return "MockActivity"; }
+  std::string ActivityDebugTag(WakeupMask /*mask*/) const override {
+    return DebugTag();
+  }
+
+  void Activate() {
+    if (scoped_activity_ == nullptr) {
+      scoped_activity_ = std::make_unique<ScopedActivity>(this);
+    }
+  }
+
+  void Deactivate() { scoped_activity_.reset(); }
+
+ private:
+  std::unique_ptr<ScopedActivity> scoped_activity_;
+};
+
+MATCHER(IsPending, "") {
+  if (arg.ready()) {
+    *result_listener << "is ready";
+    return false;
+  }
+  return true;
+}
+
+MATCHER(IsReady, "") {
+  if (arg.pending()) {
+    *result_listener << "is pending";
+    return false;
+  }
+  return true;
+}
+
+MATCHER_P(IsReady, value, "") {
+  if (arg.pending()) {
+    *result_listener << "is pending";
+    return false;
+  }
+  if (arg.value() != value) {
+    *result_listener << "is " << ::testing::PrintToString(arg.value());
+    return false;
+  }
+  return true;
+}
+
+}  // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+// Layout
+
+namespace filters_detail {
+
+TEST(LayoutTest, Empty) {
+  Layout<FallibleOperator<ClientMetadataHandle>> l;
+  EXPECT_EQ(l.ops.size(), 0u);
+  EXPECT_EQ(l.promise_size, 0u);
+  EXPECT_EQ(l.promise_alignment, 0u);
+}
+
+TEST(LayoutTest, Add) {
+  Layout<FallibleOperator<ClientMetadataHandle>> l;
+  l.Add(1, 4,
+        FallibleOperator<ClientMetadataHandle>{&l, 120, nullptr, nullptr,
+                                               nullptr});
+  EXPECT_EQ(l.ops.size(), 1u);
+  EXPECT_EQ(l.promise_size, 1u);
+  EXPECT_EQ(l.promise_alignment, 4u);
+  EXPECT_EQ(l.ops[0].call_offset, 120);
```

Comment:
ack.

There's a separate test for size everywhere, so I think we're covered there - and I tend to find this form is more readable - I find it very difficult to determine what's actually getting tested once things like `testing::Field` are in play.

Reply:
Maybe something like the following would be more readable?

```
MATCHER_P2(OperatorEq, channel_data, call_offset, "") {
  bool ok = true;
  ok &= ::testing::ExplainMatchResult(channel_data, arg.channel_data, result_listener);
  ok &= ::testing::ExplainMatchResult(call_offset, arg.call_offset, result_listener);
  return ok;
}

using ::testing::_;
using ::testing::ElementsAre;

EXPECT_THAT(l.ops, ElementsAre(OperatorEq(_, 120)));
```

If you leave this as-is, it would probably be a good idea to at least change the size check from an EXPECT to an ASSERT, so that the test won't crash if it's the wrong size.